File: opcua-webservice-spring/examples/uawebservicetest/src/main/java/org/opcfoundation/uawebservicetest/testdigitaltwin/DigitalTwinTestType.java
package org.opcfoundation.uawebservicetest.testdigitaltwin;

import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.digitaltwin.DigitalTwinSpace;
import org.opcfoundation.webserver.digitaltwin.digitaltwin.DigitalTwinType;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDescriptorRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDescriptorResponse;
import org.opcfoundation.webserver.service.message.digitaltwin.GetSubmodelsRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetSubmodelsResponse;
import org.opcfoundation.webserver.types.digitaltwin.SubmodelDescriptor;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public class DigitalTwinTestType extends DigitalTwinType {
    public DigitalTwinTestType(
            SubmodelTestType submodel,
            DynamicSubmodelTestType elementListSubmodel,
            DigitalTwinSpace space) {
        super("TestDigitalTwin", new LocalizedText("TestDigitalTwin"), space);
        setDescription(new LocalizedText("TestDigitalTwin"));

        addSubmodel(
                submodel,
                "Submodel",
                new LocalizedText("Submodel"),
                new LocalizedText("Test Submodel"));

        addSubmodel(
                elementListSubmodel,
                "ElementListSubmodel",
                new LocalizedText("ElementListSubmodel"),
                new LocalizedText("Test ElementListSubmodel"));
    }

    public CompletableFuture<GetDescriptorResponse> onGetDescriptor(GetDescriptorRequest request)
    {
        GetDescriptorResponse response = new GetDescriptorResponse(
                new LocalizedText("DT " + request.getId()),
                new LocalizedText("Test Digital Twin number " + request.getId()));

        return CompletableFuture.completedFuture(response);
    }

    public CompletableFuture<GetSubmodelsResponse> onGetSubmodels(GetSubmodelsRequest request) {
        List<UaObject> submodels = getSubmodels();
        GetSubmodelsResponse response = new GetSubmodelsResponse();

        String id = request.getId();
        if (id.equals("5"))
        {
            id = id + "-16";
        }

        for (UaObject item : submodels)
        {
            if (id.equals("4") && item.browseName().equals("ElementListSubmodel")) continue;
            response.add(new SubmodelDescriptor(id, item));
        }

        if (id.equals("3")) {
            response.add(new SubmodelDescriptor(id, new LocalizedText("Submodel-" + id), DigitalTwinSpaceTest.submodelTestType));
        }

        return CompletableFuture.completedFuture(response);
    }
}

File: opcua-webservice-node/examples/testserver/testdigitaltwin/DigitalTwinTestType.ts
import { NodeClass } from "opcua-webapi";
import { UaLocalizedText } from "opcua-webapi-ts";
import { UaObject } from "../../../src/server/addressspace/nodes";
import { DigitalTwinSpace } from "../../../src/server/digitaltwin";
import { DigitalTwinType } from "../../../src/server/digitaltwin/digitaltwin/DigitalTwinType";
import { GetDescriptorRequest, GetDescriptorResponse, GetSubmodelsRequest, GetSubmodelsResponse } from "../../../src/server/service/message";
import { SubmodelDescriptor } from "../../../src/server/types/digitaltwin/SubmodelDescriptor";
import { DigitalTwinSpaceTest } from "./DigitalTwinSpaceTest";
import { DynamicSubmodelTestType } from "./DynamicSubmodelTestType";
import { SubmodelTestType } from "./SubmodelTestType";

export class DigitalTwinTestType extends DigitalTwinType {
    constructor(
        submodel: SubmodelTestType,
        elementListSubmodel: DynamicSubmodelTestType,
        space: DigitalTwinSpace,
    ) {
        super("TestDigitalTwin", new UaLocalizedText("TestDigitalTwin"), space);
        this.description = new UaLocalizedText("TestDigitalTwin");

        this.addSubmodel(
            submodel,
            "Submodel",
            new UaLocalizedText("Submodel"),
            new UaLocalizedText("Test Submodel"),
        );

        this.addSubmodel(
            elementListSubmodel,
            "ElementListSubmodel",
            UaLocalizedText.from("ElementListSubmodel"),
            UaLocalizedText.from("Test ElementListSubmodel"),
        );
    }

    override async onGetDescriptor(request: GetDescriptorRequest): Promise<GetDescriptorResponse> {
        return new GetDescriptorResponse(
            new UaLocalizedText(`DT ${request.id}`),
            new UaLocalizedText(`Test Digital Twin number ${request.id}`),
        );
    }

    override async onGetSubmodels(request: GetSubmodelsRequest): Promise<GetSubmodelsResponse> {
        const response = new GetSubmodelsResponse();
        const submodels = this.getMembers().filter((item) => item.nodeClass === NodeClass.Object) as Array<UaObject>;

        let id = request.id;
        if (id === "5") {
            id = `${id}-16`;
        }

        for (const item of submodels) {
            if (id === "4" && item.browseName === "ElementListSubmodel") {
                continue;
            }

            response.add(new SubmodelDescriptor(id, item));
        }

        if (id === "3") {
            response.add(new SubmodelDescriptor(id, new UaLocalizedText(`Submodel-${id}`), DigitalTwinSpaceTest.submodelTestType));
        }

        return response;
    }
}